### PR TITLE
WIP: Creating Metadata Artifact

### PR DESCRIPTION
**Pull request by mortonjt**
This is something that I pair programmed with @ebolyen .  This is still work in progress.

### CODE REVIEW

**Comment by ebolyen** [comment]
You lost `test_util.py` because it was a new file.

File: qiime2/core/testing/transformer.py [hunk 1]
```diff
@@ -88,6 +89,9 @@ def _14(ff: MappingFormat) -> dict:
             data[key] = value
     return data
 
+@dummy_plugin.register_transformer
+def _134(df: MappingDirectoryFormat) -> Metadata:
```
Comment (thermokarst) [inline qiime2/core/testing/transformer.py]: 👍, sweet function name!

File: qiime2/core/archive/provenance.py [hunk 2]
```diff
@@ -269,10 +269,15 @@ def handle_metadata(self, name, value):
         else:
             raise NotImplementedError
 
+        _uuid = ""
```
Comment (ebolyen) [inline qiime2/core/archive/provenance.py]: `_uuid` -> `uuid_ref`

Comment (ebolyen) [inline qiime2/core/archive/provenance.py]: (Also usually you put the `_` on the end of the variable if you're trying not to clobber a builtin, but I think adding `ref` will make it clearer)

File: qiime2/metadata.py [hunk 3]
```diff
@@ -15,13 +15,43 @@
...
+        Parameters
+        ----------
+        artifact: qiime2.Artifact
+           Loaded artifact object.
+
+        Returns
+        -------
+        qiime2.Metadata
+        """
+        import qiime2.sdk.util
```
Comment (ebolyen) [inline qiime2/metadata.py]: Move this import to top of file

File: qiime2/metadata.py [hunk 4]
```diff
@@ -46,11 +76,14 @@ def get_category(self, *names):
                                       " supported.")
         try:
             result = MetadataCategory(self._dataframe[names[0]])
+
```
Comment (ebolyen) [inline qiime2/metadata.py]: drop extra line

File: qiime2/sdk/util.py [hunk 5]
```diff
@@ -92,3 +92,25 @@ def parse_format(format_str):
     except KeyError:
         raise TypeError("No format: %s" % format_str)
     return format_record.format
+
+
+def has_metadata(artifact):
```
Comment (ebolyen) [inline qiime2/sdk/util.py]: Do we want to push this into the `Artifact` object's API? I'm fine with leaving in sdk for now and moving later on. 

pinging @jairideout @thermokarst

File: qiime2/sdk/util.py [hunk 6]
```diff
@@ -92,3 +92,25 @@ def parse_format(format_str):
...
+       The artifact to check for metadata.
+
+    Returns
+    -------
+    bool
+       True if the artifact has metadata, False otherwise.
+
+    """
+    from qiime2 import Metadata
+    from qiime2.core import transform
```
Comment (ebolyen) [inline qiime2/sdk/util.py]: move import up

Comment (ebolyen) [inline qiime2/sdk/util.py]: change to fully qualified imports as well. (We tend to avoid `from` in the framework)

File: qiime2/sdk/util.py (hunk 5 above)
Comment (jairideout) [inline qiime2/sdk/util.py]: I'm in favoring of moving it now (see comment [here](https://github.com/qiime2/q2cli/pull/137#discussion_r115790323))

**Comment by ebolyen** [comment]
Oops! Looks like I broke a test by changing the transformer's generated index to a string.

File: qiime2/sdk/tests/test_util.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# ----------------------------------------------------------------------------
```
Comment (jairideout) [inline qiime2/sdk/tests/test_util.py]: Looks like this file is left over from a refactor, can you remove it?

File: qiime2/tests/test_metadata.py [hunk 8]
```diff
@@ -36,6 +37,30 @@ def test_does_not_cast_index_or_column_types(self):
...
+        with self.assertRaisesRegex(ValueError, 'metadata'):
+            qiime2.Metadata.from_artifact(A)
+
+    def test_artifact(self):
+        A = qiime2.Artifact.import_data('Mapping', {'a': ['1', '2'],
+                                                    'b': ['2', '3']})
+        md = qiime2.Metadata.from_artifact(A)
+        art = md.artifact
+        self.assertIsInstance(art, qiime2.Artifact)
+        self.assertEquals(art, A)
```
Comment (jairideout) [inline qiime2/tests/test_metadata.py]: Can you use `self.assertEqual`? `assertEquals` is [deprecated](https://docs.python.org/3/library/unittest.html#deprecated-aliases).

**Comment by ebolyen** [comment]
Good catches, thanks @jairideout! Should be fixed now.

**Comment by jairideout** [comment]
👍 ! Can you fix the merge conflicts plz?

**Comment by ebolyen** [comment]
Oh no!

**Pull request merged**